### ISSUE TITLE
Serializer-specific struct serialization

### ISSUE THREAD

**Issue by rubdos** [serde-rs/serde#968]
Hello,

(**TL;DR** at the bottom)

### Backstory

I am currently trying to implement an ASN.1 serialization system for serde. ASN.1 DER (and very possibly other serializations) have a specific way of serializing different types; for example, an integer is prepended by a length, then encoded using two's complement, or Date types are encoded in a Zulu notation in ASCII (which I find weird for a binary notation, but that's besides the point).

Now, in my example, I need to be able to serialize longer integer types than u64; in fact, I need infinite length integers, such as the `BigInt` type provided by the [`num`](/rust-num/num) crate. I already can serialize it internally, but binding it to serde means that `BigInt` types in structs get serialized through their own serialization, as specified by the `num` crate.

Same goes for date-time types.

### Currently possible solution 1

As @sfackler mentioned on IRC, it is possible to add a helper `trait` to the target crate (e.g., add a `trait BigIntSerializer` to `num`), for which you make a default impl, and you invoke it's `serialize` method from the `impl serde::Serialize for BigInt` implementation:

```Rust
impl<T> BigIntSerializer for T
    where T: serde::Serializer
    type Ok = T::Ok;
    type Error = T::Error;
    default fn serialize_bigint(self, data: &BigInt) -> Result<<Self as SerializeBigInt>::Ok, Self::Error> {
        use serde::Serialize;
        (data.sign, &data.data).serialize(self)
    }
```

(something like that)

After that, one can optionally specialize this `trait` in e.g. the ASN.1 crate to account for the more detailed serialization (which then can also be done without overhead).

N.b., this approach obviously depends on Rust's unstable [specialization feature](https://github.com/rust-lang/rust/issues/31844).

Now, this approach requires an effort for all types (and a major change in their crates!) that are to be serialized. The upside is that it retains all API compatibility.

### Currently possible solution 2

```Rust
#[derive(Serialize)]
struct MyType {
    #[serde(with="customASN1BigIntSerializer")]
    p: BigInt,
}
```

but this is not scalable, especially since I want it to be Serializer specific, and I don't really want to write a serializer that dispatches on the Serializer type :-)

## TL;DR

I'm wondering whether it would be possible to have serde provide an interface to "override" the default `Serialize` depending on the Serializer.

### Future possible solution?

If we move the generic `S: Serialize` type to the level of the `Serialize` type, instead of its `serialize` method, I would suspect that it would be easy to have the kind of specialization, but apparently [there's a bug with that](https://github.com/rust-lang/rust/issues/38516), but it remains unclear to me what that bug is about. FWIW, it didn't compile for me this week.

A second issue with that approach would be that the API gets broken (again, right after everyone moved to serde 1.0), and that there would be a `#[cfg]`-hell until specialization gets in stable (which will probably take a while).

---

So, my question: what would the serde authors propose for this?

**Comment by dtolnay** [comment]
Look into how the toml crate handles datetimes. The `toml::Value` enum has a Datetime member that serializes and deserializes seamlessly as toml's native datetime representation. The solution we have there is more flexible and more correct than either specialization or Serialize\<S>.

**Comment by rubdos** [comment]
Seems to me like [they dispatch on the struct name](https://github.com/alexcrichton/toml-rs/blob/master/src/ser.rs#L554) in `serialize_struct`, which is dispatch on runtime, while (I think) all information is actually already available on compile time. They also had to impl a complete Serializer just for the Datetime.

I do have a `Asn1::GeneralizedTime(DateTime)` member that serializes to Asn1's native type (same for a BigInt); it's just that I have to actually get the struct from serde.

**Comment by vojtechkral** [comment]
For the benefit of whoever comes across this problem in the furutre: The above link to toml is broken, I think it was supposed to reference [this line](https://github.com/alexcrichton/toml-rs/blob/022e914ccf9ae0ed61e5fd09152b41f37737602a/src/ser.rs#L945) in toml-rs.
